### PR TITLE
migrate Mocha's tests to Unexpected assertion library

### PR DESCRIPTION
**Pull request by boneskull**
These changes convert (nearly) all usages of [expect.js](https://npm.im/expect.js) and [assert](https://npm.im/assert) to [unexpected](http://unexpected.js.org).

Motivation:

1. Consistency
2. Better diffs
3. Custom assertions (see the new `test/assertions.js`) are easy to implement
4. Easier to avoid potentially throwing multiple errors in a single test w/ multiple assertions (see `to satisfy`)
5. Out-of-the-box support for `Promise`
6. Active development

- I removed the userland `assert` b/c as I understand the Node.js team is no longer discouraging the use of the built-in module.
- `chai` is added because we have a test that asserts compatibility, which was previously not-so-valuable, because we weren't actually testing against Chai itself.  Whether or not we should remove this particular test is debatable.
- We now have custom assertions for handling the result of `runMochaJSON` in the integration tests
- `test/unit/runnable.spec.js` uses *no* assertion library.  See my comments there and in #3342.
- There's now a `test/README.md` that explains some things.

### CODE REVIEW

**Comment by boneskull** [comment]
should probably fix those windows-only tests.

**Comment by boneskull** [comment]
I’m not too concerned about it creating friction. Even if it did, I feel the benefits of using it outweigh any negatives about the general unfamiliarity with it. 

What we should do, however, is document the custom assertions and make it clear where the documentation for the base assertions live. If a contributor is familiar with Chai, it shouldn’t be too jarring, anyway. 

That said, to someone unfamiliar with *either*, I’d argue Unexpected is fundamentally easier to understand than Chai, since it doesn’t do strange things with object properties:

```js
// chai
// this is really weird if you’ve never seen it before
expect(foo).to.be.ok; 

// unexpected
expect(foo, ‘to be ok’);
```

**Comment by Munter** [comment]
I'm obviously all for this change. But I'm completely biased, having contributed to the unexpected ecosystem myself.

**Comment by plroebuck** [comment]
Had never heard or even seen this of this library; Google searches for `expect` all reference "chai".
Really dislike the use of 'to be' comparing two different variables; I immediately think of comparing their addresses, not values. That said, glancing the code changes, I notice several places where the new assertion should probably be different. 

For example, TAP reporter spec had this:
     `expect(stdout[0], 'to equal', expectedMessage);`
should probably be:
     `expect(stdout[0], 'to be', expectedMessage);`

Similar elsewhere. Sorry, but couldn't comment inline.

**Comment by boneskull** [comment]
`to be` is a strict equality which is going to be akin to a reference.

**Comment by boneskull** [comment]
and yes, we should probably audit those assertions as this was mostly a search/replace job.

**Comment by boneskull** [comment]
but if it ended up being `to equal` it was originally `.to.eql`. in this case it was already incorrect

**Comment by boneskull** [comment]
@plroebuck curious what the value of the Google search is?  `expect` is a variable. I could call it `fireball` instead.

File: test/integration/options.spec.js [hunk 1]
```diff
@@ -75,12 +62,8 @@ describe('options', function () {
           done(err);
           return;
         }
-        assert.equal(res.stats.pending, 0);
-        assert.equal(res.stats.passes, 0);
-        assert.equal(res.stats.failures, 1);
-
-        assert.equal(res.failures[0].title, 'should only display this error');
-        assert.equal(res.code, 1);
+        expect(res, 'to have failed')
```
[the remark this answers is not in the document]
Comment (boneskull) [reply test/integration/options.spec.js]: good catch. note that this implies “to have failed” but `to have passed test count` does **not**

[the remark this answers is not in the document]
Comment (boneskull) [reply test/integration/options.spec.js]: (imply `to have passed`). sorry on phone

File: test/assertions.js [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+      expect,
+      result,
+      count
+    ) {
+      expect(result.stats.tests, '[not] to be', count);
+    })
+    .addAssertion(
+      '<JSONResult> [not] to have failed [test] count <number>',
+      function (expect, result, count) {
+        expect(result.stats.failures, '[not] to be', count);
```
[the remark this answers is not in the document]
Comment (boneskull) [reply test/assertions.js]: yeah, hmmm... so what I didn’t do was create customized diffs for each assertion. you can do this.  but didn’t see it as necessary for a first pass. 

what you’re saying comes in to play when working with the diffs—I’m not sure which of the two forms will more easily provide a better diff once it’s customized.

[the remark this answers is not in the document]
Comment (outsideris) [reply test/assertions.js]: Agreed. I'm interested in the diff, but I'm also not sure which is better.

**Comment by plroebuck** [comment]
> _to be_ is a strict equality which is going to be akin to a reference.

Unexpected's documentation says _to be_ uses [SameValue](http://ecma-international.org/ecma-262/5.1/#sec-9.12) equality, which is different from `===`.

**Comment by boneskull** [comment]
@plroebuck you're right. though the documentation page @ ECMA is suspended, apparently?

Either way, their implementation makes more sense, since things like `NaN === NaN` evaluate to `false`. thanks javascript

**Comment by boneskull** [comment]
squashed and rebased and reformatted; @outsideris' comment addressed in 03b3c4a

**Comment by boneskull** [comment]
weird, how did `options.spec.js` get reverted...uarggghh

**Comment by plroebuck** [comment]
I know this is already merged, but...

<https://github.com/mochajs/mocha/blob/85c8f0dc173d76802ad97ea642220b251d4fe5cd/test/node-unit/file-utils.spec.js#L44-L59>

I fail to see _any_ value to using `expectedLength` here, since it's _really_ being used to calculate the actual length. #L58 will _always_ be correct if the other two tests pass.

Should remove all lines containing `expectedLength`. As a better fix, collect the expected files in an array and spread them in single call to `expect()`. Something like:

```js
    it('should accept a glob "path" value', function() {
      var res = utils
        .lookupFiles(tmpFile('mocha-utils*'), ['js'], false)
        .map(path.normalize.bind(path));

      var expectedFiles = [
        tmpFile('mocha-utils.js')
      ];
      if (symlinkSupported) {
        expectedFiles.push(tmpFile('mocha-utils-link.js'));
      }
      expect(res, 'to contain', ...expectedFiles);
    });
```

**Pull request merged**